src/DurableTask.AzureStorage/Storage/BlobContainer.cs

```diff
@@ -68,9 +70,12 @@ public async Task<bool> DeleteIfExistsAsync(string? appLeaseId = null, Cancellat
 
         public AsyncPageable<Blob> ListBlobsAsync(string? prefix = null, CancellationToken cancellationToken = default)
         {
-            return new AsyncPageableProjection<BlobItem, Blob>(
-                this.blobContainerClient.GetBlobsAsync(BlobTraits.Metadata, BlobStates.None, prefix, cancellationToken),
-                x => this.GetBlobReference(x.Name)).DecorateFailure();
+            return this.blobContainerClient
+                .GetBlobsAsync(BlobTraits.Metadata, BlobStates.None, prefix, cancellationToken)
+                .DecorateFailure()
+                .TransformPages(p => p.Values
+                    .Where(b => !IsHnsFolder(b))
+                    .Select(b => this.GetBlobReference(b.Name)));
```

Comment:
nit - let's add an inline comment here documenting why we have the `IsHnsFolder` filter.
```suggestion
            //  When hierarchical namespaces (HNS) are enabled for a storage account, the "List Blobs" API returns the
            // folder itself as a "blob" in the response (in addition to the blobs themselves), 
            // which can result in a NullReferenceException if not accounted for. 
            // Therefore, we filter out all HNS folders in the result
            return this.blobContainerClient
                .GetBlobsAsync(BlobTraits.Metadata, BlobStates.None, prefix, cancellationToken)
                .DecorateFailure()
                .TransformPages(p => p.Values
                    .Where(b => !IsHnsFolder(b))
                    .Select(b => this.GetBlobReference(b.Name)));
```